test/object-store/sync/flx_sync.cpp

```diff
@@ -4995,6 +4967,464 @@ TEST_CASE("flx: nested collections in mixed", "[sync][flx][baas]") {
+    };
+
+    SECTION("Role changes lead to objects in/out of view without client reset") {
+        FLXSyncTestHarness harness("flx_role_change_bootstrap", {g_person_schema, {"role", "firstName", "lastName"}});
+        // Get the current rules so it can be updated during the test
+        auto& app_session = harness.session().app_session();
+        auto test_rules = app_session.admin_api.get_default_rule(app_session.server_app_id);
+
+        SECTION("Basic bootstraps") {
+            // Single message bootstrap - remove employees, keep mgrs/dirs
+            // 5000 emps, 100 mgrs, 25 dirs
+            // num_objects_before_bootstrap_flush: 10
+            TestParams params{5000, 100, 25, 10};
+            auto num_total = params.num_emps + params.num_mgrs + params.num_dirs;
+            auto realm = setup_test(harness, params, {}, num_total);
+
+            // Single message bootstrap - remove employees, keep mgrs/dirs
+            logger->trace("ROLE CHANGE: Updating rules to remove employees");
+            update_role(test_rules, {{"role", {{"$in", {"manager", "director"}}}}});
+            update_perms_and_verify(harness, realm, test_rules,
+                                    {BootstrapMode::SingleMessage, 0, params.num_mgrs, params.num_dirs});
+            // Write the same rules again - the client should not receive the reconnect (200) error
+            logger->trace("ROLE CHANGE: Updating same rules again and verify reconnect doesn't happen");
+            update_perms_and_verify(harness, realm, test_rules,
+                                    {BootstrapMode::NoReconnect, 0, params.num_mgrs, params.num_dirs});
+            // Multi-message bootstrap - add employeees, remove managers and directors
+            logger->trace("ROLE CHANGE: Updating rules to add back the employees and remove mgrs/dirs");
+            update_role(test_rules, {{"role", "employee"}});
+            update_perms_and_verify(harness, realm, test_rules, {BootstrapMode::MultiMessage, params.num_emps, 0, 0});
+        }
+        SECTION("Single-message/Multi-changeset bootstrap") {
+            /// TODO: Update this test to use the PAUSE_DOWNLOAD_BUILDER test command
+            ///       once it can be sent prior to the IDENT message
+            // 500 emps, 500 mgrs, 125 dirs
+            TestParams params{500, 500, 125, 100, 2000};
+            update_role(test_rules, {{"role", "employee"}});
+            auto realm = setup_test(harness, params, test_rules, params.num_emps);
+
+            // Single message/multi-changeset bootstrap - add back managers and directors
+            logger->trace("ROLE CHANGE: Updating rules to all records");
+            update_role(test_rules, true);
+            update_perms_and_verify(
+                harness, realm, test_rules,
+                {BootstrapMode::SingleMessageMulti, params.num_emps, params.num_mgrs, params.num_dirs});
+        }
+    }
+    SECTION("Role changes for one user do not change unaffected user") {
```

Comment:
is this something we need to test? seems like something the server should have tests for.

Reply:
sounds like the server doesn't have tests for this...